src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -395,24 +465,63 @@ private bool TestStream(Stream stream)
             return false;
         }
 
-        private void RegisterHandler(Guid runtimeId, StreamHandler handler)
+        private void ProvideEndpointInfo(in IpcEndpointInfo endpointInfo)
         {
-            lock (_lock)
+            lock (_newEndpointInfoLock)
             {
-                if (!_cachedStreams.TryGetValue(runtimeId, out Stream stream))
+                IpcEndpointInfo? endpointInfoNullable = endpointInfo;
+                // Provide the endpoint info to the first handler that will accept it.
+                for (int i = 0; i < _newEndpointInfoHandlers.Count && endpointInfoNullable.HasValue; i++)
                 {
-                    throw new InvalidOperationException($"Runtime instance with identifier '{runtimeId}' is not registered.");
+                    EndpointInfoHandler handler = _newEndpointInfoHandlers[i];
+                    // Handler will return true if it has completed
+                    if (handler(ref endpointInfoNullable))
+                    {
+                        _newEndpointInfoHandlers.RemoveAt(i);
```

Comment:
The comments suggest we should break the loop here?

Reply:
The for-condition breaks the loop when the endpointInfoNullable is nulled.

Reply:
Ah I missed that.

Reply:
This more obvious with the explicit parameter stating that the value was consumed.